File: SharedCore/Adapters/Swift/SharedCoreTrafficDisplayFormatter.swift
import Foundation

enum SharedCoreTrafficDisplayFormatter {
  enum Error: Swift.Error, Equatable {
    case unsupportedDecimalPlaces(UInt32)
  }

  static func string(
    from scale: SharedTrafficScale,
    format: SharedCoreTrafficFormat
  ) throws -> String {
    let number = try formattedNumber(from: scale)
    let unit = unitText(for: scale.unit, format: format)

    switch format {
    case .byteCount:
      return "\(number) \(unit)"
    case .rate:
      return "\(number) \(unit)/s"
    case .compactRate:
      return "\(number)\(unit)/s"
    }
  }

  private static func formattedNumber(from scale: SharedTrafficScale) throws -> String {
    let format: String
    switch scale.decimalPlaces {
    case 0:
      format = "%.0f"
    case 1:
      format = "%.1f"
    case 2:
      format = "%.2f"
    default:
      throw Error.unsupportedDecimalPlaces(scale.decimalPlaces)
    }
    return String(
      format: format,
      locale: Locale(identifier: "en_US_POSIX"),
      scale.value
    )
  }

  private static func unitText(
    for unit: SharedTrafficUnit,
    format: SharedCoreTrafficFormat
  ) -> String {
    switch (unit, format) {
    case (.bytes, _):
      return "B"
    case (.kilobytes, .compactRate):
      return "K"
    case (.megabytes, .compactRate):
      return "M"
    case (.gigabytes, .compactRate):
      return "G"
    case (.terabytes, .compactRate):
      return "T"
    case (.kilobytes, _):
      return "KB"
    case (.megabytes, _):
      return "MB"
    case (.gigabytes, _):
      return "GB"
    case (.terabytes, _):
      return "TB"
    }
  }
}

File: platform/windows/MihomoMeter.Windows.Core/Application/SharedCoreTrafficDisplayFormatter.cs
using System.Globalization;

namespace MihomoMeter.Windows.Core.Application;

internal static class SharedCoreTrafficDisplayFormatter
{
    public static string Format(
        SharedTrafficScale scale,
        SharedCoreTrafficFormat format)
    {
        var numberFormat = scale.DecimalPlaces switch
        {
            0 => "0",
            1 => "0.0",
            2 => "0.00",
            _ => throw new InvalidOperationException(
                $"共享核心返回了不支持的小数位数：{scale.DecimalPlaces}。"),
        };
        var number = scale.Value.ToString(numberFormat, CultureInfo.InvariantCulture);
        var unit = UnitText(scale.Unit, format);
        return format switch
        {
            SharedCoreTrafficFormat.ByteCount => $"{number} {unit}",
            SharedCoreTrafficFormat.Rate => $"{number} {unit}/s",
            SharedCoreTrafficFormat.CompactRate => $"{number}{unit}/s",
            _ => throw new InvalidOperationException($"未知共享核心流量格式：{format}。"),
        };
    }

    private static string UnitText(
        SharedTrafficUnit unit,
        SharedCoreTrafficFormat format)
    {
        if (format == SharedCoreTrafficFormat.CompactRate)
        {
            return unit switch
            {
                SharedTrafficUnit.Bytes => "B",
                SharedTrafficUnit.Kilobytes => "K",
                SharedTrafficUnit.Megabytes => "M",
                SharedTrafficUnit.Gigabytes => "G",
                SharedTrafficUnit.Terabytes => "T",
                _ => throw new InvalidOperationException(
                    $"共享核心返回未知流量单位：{unit}。"),
            };
        }

        return unit switch
        {
            SharedTrafficUnit.Bytes => "B",
            SharedTrafficUnit.Kilobytes => "KB",
            SharedTrafficUnit.Megabytes => "MB",
            SharedTrafficUnit.Gigabytes => "GB",
            SharedTrafficUnit.Terabytes => "TB",
            _ => throw new InvalidOperationException($"共享核心返回未知流量单位：{unit}。"),
        };
    }
}
